libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Properties.hs

```diff
@@ -432,38 +430,94 @@ genTxOut proof val = do
+  -- Note: Never add an old pair unless there are at least 1 new pairs
+  let maybeCons (Just pair) xs | length xs > 1 = pair : xs
+      maybeCons _ xs = xs
+  pure (UTxO (SplitMap.fromList (maybeCons maybepair pairs)))
   where
     genOut = do
       val <- lift genPositiveVal
       fields <- genTxOut reify val
       pure (coreTxOut reify fields)
 
-genDCert :: Reflect era => GenRS era (DCert (Crypto era))
+-- | Generate both the spending and reference inputs and a key from the spending
+--   inputs we can use to pay the fee. That key is never from the oldUTxO
+genSpendReferenceInputs ::
+  Proof era ->
+  Map (TxIn (Crypto era)) (Core.TxOut era) ->
+  GenR
+    era
+    ( TxIn (Crypto era), -- The feekey, used to pay the fee.
+      Map (TxIn (Crypto era)) (Core.TxOut era),
+      Map (TxIn (Crypto era)) (Core.TxOut era)
+    )
+genSpendReferenceInputs proof newUTxO = do
+  let pairs = SplitMap.toList newUTxO
+  numInputs <- lift $ choose (1, length pairs)
+  numRefInputs <- lift $ choose (0, maxRefInputs proof)
+  badTest <- getPrevUtxoTest
+  ((feeinput, _), inputPairs) <- lift $ chooseGood (badTest . fst) numInputs pairs
+  refInputPairs <- take numRefInputs <$> lift (shuffle $ pairs)
+  pure (feeinput, Map.fromList inputPairs, Map.fromList refInputPairs)
+
+-- | 'xs' is a non empty list. It has at most one 'bad' element
+--   if 'xs' has length 1, the element is guarantied to be good.
+--   'n' is a positive Int, ranging between (1 .. length xs). Return
+--   a pair (good,ys) where (bad good)==False, (good `elem` ys),
+--   (length ys) == n, and all (`elem` xs) ys. This is used to generate
+--   the spending inputs, which always contain on Input we can use
+--   to pay the fee, that does not come from the oldUTxO.
+chooseGood :: (a -> Bool) -> Int -> [a] -> Gen (a, [a])
+chooseGood bad n xs = do
+  let (good, others) =
+        case xs of
+          [] -> error "empty list is chooseGood, should never happen"
+          [x] -> (x, [])
+          (x : y : more) -> if bad x then (y, (x : more)) else (x, y : more)
+  tailx <- take (n - 1) <$> shuffle others
+  result <- shuffle (good : tailx)
+  pure (good, result)
+
+-- =====================================================
+
+genDCert :: Reflect era => GenR era (DCert (Crypto era))
 genDCert =
   elementsT
     [ DCertDeleg
         <$> elementsT
-          [ RegKey <$> genCredential Cert,
-            DeRegKey <$> genCredential Cert,
+          [ RegKey <$> genCredentialY Cert,
+            DeRegKey <$> genCredentialY Cert,
             Delegate <$> genDelegation
           ]
     ]
   where
     genDelegation = do
-      rewardAccount <- genCredential Cert
-      poolId <- genPool
+      rewardAccount <- genCredentialY Cert
+      poolId <- genPoolY
       pure $ Delegation {_delegator = rewardAccount, _delegatee = poolId}
 
-genDCerts :: Reflect era => GenRS era [DCert (Crypto era)]
+hasKey :: Ord k => k -> Map k v -> Bool
+hasKey k m = case Map.lookup k m of Just _ -> True; _ -> False
```

Comment:
These one line case statements feel really unnatural :sob: 

Regardless of that: `hasKey` is just `Map.member`.

```suggestion
```

Reply:
Good call, Map.member is is!